### PR TITLE
Fix issue of a subset of block captured by transform-feedback buffer

### PR DESCRIPTION
**Pull request by xuechen417**
Current spir-v reader cannot handle that case that a subset block
members are captured by transform-feedback. It forces the whole block to
bbe captured. This change will add the support for the case in two
aspects:
1. consider the block memeber decoration for generating `!lgc.xfb.state`
2. correct the setting of `ShaderInOutMetadata` to generate
   `lgc.create.write.xfb.output` only for the captured data

PS: the list test is changed to .spvasm because current glslang don't
   support this case either.

Fixes:[Angle]dEQP-GLES3.functional.transform_feedback.point_size.points_separate

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5432bf58a42da2786186ffb5b4453387d10e333d
<details><summary>CTS tests (Failed: 0/211487)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39248/70495 (55.7%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 31247/70495 (44.3%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
This is one heck of a weird corner case. Thank you for adding the test :)

That said, only `gl_PointSize` should be streamed out, but I see two calls for `write.xfb.output` in the test. I think there should only be one for gl_PointSize, the one for gl_Position needs to be removed.

**Comment by xuechen417** [comment]
> This is one heck of a weird corner case. Thank you for adding the test :)
> 
> That said, only `gl_PointSize` should be streamed out, but I see two calls for `write.xfb.output` in the test. I think there should only be one for gl_PointSize, the one for gl_Position needs to be removed.

Thank you for catching it. This seems a legacy issue. I will take a look at it.

**Comment by xuechen417** [comment]
V1: Fix the issue of creating `lgc.create.write.xfb.output` for non-captured output.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4369cb2a9005305dd28a2f462261dac9cd2c4487
<details><summary>CTS tests (Failed: 0/211487)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39248/70495 (55.7%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 31247/70495 (44.3%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/test/shaderdb/core/TestXfbStateMetadata.spvasm [hunk 1]
```diff
@@ -0,0 +1,109 @@
+; BEGIN_SHADERTEST
+; RUN: amdllpc -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: {{^// LLPC}} SPIRV-to-LLVM translation results
+; SHADERTEST: define dllexport spir_func void @main() #2 !spirv.ExecutionModel !6 !lgc.xfb.state !7 !lgc.shaderstage !8
+; SHADERTEST: !7 = !{i32 0, i32 4, i32 -1, i32 0, i32 -1, i32 0, i32 -1, i32 0}
+; SHADERTEST-LABEL: {{^// LLPC}} transform feedback export info (vertex shader)
+; SHADERTEST: float (builtin = PointSize), xfbBuffer = 0, xfbStride = 4, xfbOffset = 0
+; SHADERTEST: AMDLLPC SUCCESS
```
Comment (nhaehnle) [inline llpc/test/shaderdb/core/TestXfbStateMetadata.spvasm:8]: Add a SHADERTEST-NOT line to make sure that builtin = Position does not appear.

**Review by nhaehnle — COMMENTED** [review]
Any particular reason to switch to a SPIR-V test? The GLSL one was easier to read...

Apart from that the change looks okay.

File: llpc/test/shaderdb/core/TestXfbStateMetadata.spvasm (hunk 1 above)
Comment (xuechen417) [reply llpc/test/shaderdb/core/TestXfbStateMetadata.spvasm:8]: Done.

**Comment by xuechen417** [comment]
> Any particular reason to switch to a SPIR-V test? The GLSL one was easier to read...

Yes, glslang has a bug and it will give wrong result for this case as below:
```
               OpMemberDecorate %gl_PerVertex 0 BuiltIn Position
               OpMemberDecorate %gl_PerVertex 1 Offset 0
               OpMemberDecorate %gl_PerVertex 1 BuiltIn PointSize
               OpDecorate %gl_PerVertex Block
               OpDecorate %_ XfbBuffer 0
               OpDecorate %_ XfbStride 4
               OpDecorate %pos Location 0
               OpDecorate %pointSize Location 1
```
I will update the test to use GLSL once the issue is fixed in glslang.

**Comment by xuechen417** [comment]
V1: Add a SHADERTEST-NOT line to check Position

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5e8b19ee99abb1a54b6b915319a87a1b26e0ac09
<details><summary>CTS tests (Failed: 0/211487)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39248/70495 (55.7%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 31247/70495 (44.3%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit b55bf6dba9baa91b62bce55f7cda8ec8a5b0245e
<details><summary>CTS tests (Failed: 1/211498)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39250/70506 (55.7%)</li>
<li>Failed: 0/70506 (0.0%)</li>
<li>Not Supported: 31256/70506 (44.3%)</li><li>Warnings: 0/70506 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39497/70495 (56.0%)</li>
<li><details><summary>Failed: 1/70495 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.rasterization.conservative.overestimate.samples_1.triangles.degenerate.min
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit b55bf6dba9baa91b62bce55f7cda8ec8a5b0245e
<details><summary>CTS tests (Failed: 0/211498)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39250/70506 (55.7%)</li>
<li>Failed: 0/70506 (0.0%)</li>
<li>Not Supported: 31256/70506 (44.3%)</li><li>Warnings: 0/70506 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>
